### DISCUSSION TITLE
PyBoard 1.1 & ucryptolib

### DISCUSSION THREAD

**Discussion by MarkMakies** [micropython/micropython#15199] [STM32 / Pyboard]
After much research I have not been able to work out how to get **ucryptolib** onto my original **PyBoard 1.1**.  I understand from a post by @Roberthh in the Micropython Forum Archive that 

> ucyptolib is not included in every port. It is for instance included in the build for Pyboard D, but not for Pyboard V1.1, which seems to be the module you use.

I have searched the libraries, hoping I could install, or alternative build my first uPy firmware with this module, but I can not work out how to do this.  Advice would be much appreciated,  Mark

**Comment by MarkMakies** [comment]
Any help or comments would be appreciated.  Mark

**Reply by glenn20** [reply]
In the stm32 micropython builds, the MICROPY_PY_CRYPTOLIB #define is activated by setting the MICROPY_PY_SSL #define. The PyBoard v1.1 does not set this by default. If you want to try compiling the cryptolib module into your micropython image (without also including ssl support), add the following line to [`ports/stm32/boards/PYBV11/mpconfigboard.h`](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/PYBV11/mpconfigboard.h):

```c
#define MICROPY_PY_CRYPTOLIB                    (1)
```

If you have no experience with building your own micropython firmware, read the docs in [`ports/stm32/README.md`](https://github.com/micropython/micropython/blob/master/ports/stm32/README.md).

Caution: I have no experience with the stm32 boards, including the PyBoard devices. There may well be a good reason why MICROPY_PY_CRYPTOLIB is not enabled by default on the pyboard, including missing dependencies or resource constraints.

PS. It is now called the [`cryptolib`](https://docs.micropython.org/en/latest/library/cryptolib.html) module.

**Reply by MarkMakies** [reply]
I managed to build my first micropython firmware and got it flashed and running on my PyBoard - awesome, thank heaps.  

If I add the define you suggested to  [ports/stm32/boards/PYBV11/mpconfigboard.h](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/PYBV11/mpconfigboard.h):  I get an error: MICROPY_PY_CRYPTOLIB is already defined in [ports/stm32/mpconfigport.h](https://github.com/micropython/micropython/blob/master/ports/stm32/mpconfigport.h)

So I changed that define from (MICROPY_PY_SSL) to (1).  

The make appeared to get all the way through the build until the very end...

```
LINK build-PYBV11/firmware.elf
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-PYBV11/py/objmodule.o:(.rodata.mp_builtin_extensible_module_table+0x1c): 
undefined reference to `mp_module_cryptolib'
```

I'm assuming this is due to the missing dependencies that you mentioned.

**Reply by glenn20** [reply]
After making changes to the config headers, you may need to run 'make clean' before you run the build step. Try that and see if it helps.

**Reply by MarkMakies** [reply]
Some progress, but still fails with a different error
```
CC ../../extmod/modcryptolib.c
../../extmod/modcryptolib.c:84:5: error: unknown type name 'AES_CTX_IMPL'
   84 |     AES_CTX_IMPL ctx;
      |     ^~~~~~~~~~~~
../../extmod/modcryptolib.c: In function 'cryptolib_aes_make_new':
../../extmod/modcryptolib.c:258:5: error: implicit declaration of function 'aes_initial_set_key_impl' [-Werror=implicit-function-declaration]
  258 |     aes_initial_set_key_impl(&o->ctx, keyinfo.buf, keyinfo.len, ivinfo.buf);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
../../extmod/modcryptolib.c: In function 'aes_process':
../../extmod/modcryptolib.c:297:9: error: implicit declaration of function 'aes_final_set_key_impl' [-Werror=implicit-function-declaration]
  297 |         aes_final_set_key_impl(&self->ctx, encrypt_mode);
      |         ^~~~~~~~~~~~~~~~~~~~~~
../../extmod/modcryptolib.c:312:17: error: implicit declaration of function 'aes_process_ecb_impl' [-Werror=implicit-function-declaration]
  312 |                 aes_process_ecb_impl(&self->ctx, in, out, encrypt);
      |                 ^~~~~~~~~~~~~~~~~~~~
../../extmod/modcryptolib.c:318:13: error: implicit declaration of function 'aes_process_cbc_impl' [-Werror=implicit-function-declaration]
  318 |             aes_process_cbc_impl(&self->ctx, in_bufinfo.buf, out_buf_ptr, in_bufinfo.len, encrypt);
      |             ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:90: build-PYBV11/extmod/modcryptolib.o] Error 1

```

**Reply by glenn20** [reply]
Ah, it looks like you cant use cryptolib without one of the tls libraries. Try adding:

```
MICROPY_PY_SSL = 1
MICROPY_SSL_MBEDTLS = 1
```

to `ports/stm32/boards/PYBV11/boards/mpconfigboard.mk`, then make clean and build again.

This will turn on full tls support and select the mbedtls implementation (which is the default). I am guessing that you might run into some resource constraints on the pyb v1.1, but it is worth a try (esp. if you dont plan to actually use http overe tls, but just want the cryptolib functionality).

**Reply by MarkMakies** [reply]
It appears that the directory mbedtls, doesn't contain the header files that modcryptolib is trying to access.  The only two files in micropython/ports/stm32/mbedtls/ are mbedtls_port.c and mbedtls_config_port.h.

```
GEN build-PYBV11/genhdr/qstr.i.last
../../extmod/modcryptolib.c:64:10: fatal error: mbedtls/aes.h: No such file or directory
   64 | #include <mbedtls/aes.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
../../extmod/modhashlib.c:35:10: fatal error: mbedtls/version.h: No such file or directory
   35 | #include "mbedtls/version.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
../../extmod/modtls_mbedtls.c:43:10: fatal error: mbedtls/platform.h: No such file or directory
   43 | #include "mbedtls/platform.h"
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
And yes, I'm not intending to use http over tls.  I just need the cryptolib functionality to encrypt low rate LoRa packets.  I though it would be easy as my PicoW board just had crypto working at the other end of my comms channel.

I also tried 

```
MICROPY_PY_SSL = 0
MICROPY_SSL_MBEDTLS = 1
```
Which gave

```
mbedtls/mbedtls_port.c:33:10: fatal error: mbedtls/platform_time.h: No such file or directory
   33 | #include "mbedtls/platform_time.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Reply by glenn20** [reply]
> It appears that the directory mbedtls, doesn't contain the header files that modcryptolib is trying to access. The only two files in micropython/ports/stm32/mbedtls/ are mbedtls_port.c and mbedtls_config_port.h.

Thats strange. I would **expect** it to find the headers in `lib/mbedtls/include/mbedtls/aes.h`. Perhaps there is one more tweak required to the header/makefiles to set the include directory... I'll take a look.

**Reply by glenn20** [reply]
Hmmm... taking a quick look, this might be a dead end. I can't see a straight forward way to setup mbedtls on the PYBV11 board. It would be good to get a definitive statement on that from someone familiar with micropython on the pyboards.

**Reply by rkompass** [reply]
Could it be that a `make submodules` is missing?

**Reply by MarkMakies** [reply]
> Hmmm... taking a quick look, this might be a dead end. I can't see a straight forward way to setup mbedtls on the PYBV11 board. It would be good to get a definitive statement on that from someone familiar with micropython on the pyboards.

Thanks Glenn for your efforts trying to find a solution for me. :)
